optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
                         Talk::getDurationInMinutes);
     }
 ----
+
+[[constraintStreamsGroupingAndCollectors]]
+== Grouping and Collectors
+
+In the constraint streams framework, grouping is a mechanism to group items in a stream according to user-provider
+criteria, similar to what a `GROUP BY` SQL clause does. Additionally, some grouping operations also accept one or more
+`Collector`s, which provide various aggregation functions.
+
+For example, the following code snippet will first group all processes by the computer they run on, sum up all the
+power required by the processes on that computer using the `ConstraintCollectors.sum(...)` collector, and finally
+penalize every computer whose processes consume more power than is available.
+
+[source,java,options="nowrap"]
+----
+    private Constraint requiredCpuPowerTotal(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudProcess.class)
+                .groupBy(CloudProcess::getComputer, sum(CloudProcess::getRequiredCpuPower))
+                .penalize("requiredCpuPowerTotal",
+                        HardSoftScore.ONE_HARD,
+                        (computer, requiredCpuPower) -> requiredCpuPower - computer.getCpuPower());
+    }
+----
+
+OptaPlanner provides several collector implementations you can use in your grouping constraint streams. This chapter
+discusses the most important ones. For a complete list, refer to the
+`org.optaplanner.core.api.score.stream.ConstraintCollectors` class.
+
+You can provide your own collectors by implementing the
+`org.optaplanner.core.api.score.stream.uni.UniConstraintCollector` interface or their `Bi...`, `Tri...` counterparts.
+The following sections will focus on the collector implementations provided out of the box.
+
+=== Counting collectors
+
+Counting collectors will count all elements in a group. For example, the following use of the
+`ConstraintCollectors.count(...)` collector will give a number of items for two separate groups - one where the talks
+have unavailable speakers, and one where they don't.
+
+[source,java,options="nowrap"]
+----
+    private Constraint speakerAvailability(ConstraintFactory factory) {
+        return factory.from(Talk.class)
+                .groupBy(Talk::hasAnyUnavailableSpeaker, count())
+                .penalize("speakerAvailability",
+                        HardSoftScore.ONE_HARD,
+                        (hasUnavailableSpeaker, count) -> ...);
+    }
+----
+
+The `ConstraintCollectors.countDistinct(...)` will do the same, but it will take object equality into account. Different
+instances of objects which still `equals(...)` will therefore only count as one instance.
+
+The return value for these collectors is a 32-bit signed integer (`int`). Should that not be enough for you, please
+refer to the `countLong()` and `countDistinctLong()` variants of these collectors.
```

Comment:
Can we describe the variants consistently similar to how we do it in here?
  https://github.com/kiegroup/optaplanner/blob/master/optaplanner-docs/src/main/asciidoc/ScoreCalculation/ScoreCalculation-chapter.adoc#hardSoftScore
Variants include: uni, bi, tri.
Variants include: int, long, bigdecimal, duration.

(I don't consider countDistinct() a variant).

Reply:
I don't think we should. The variants are easy to add, and we will be adding more as we see fit. People should refer to the Java API for a complete list - duplicating this in the docs is a recipe for pain and, more importantly, obsoleteness.

Reply:
I fully agree. We can add a short section in the beginning that everything comes in Uni/Bi/Tri etc variants later, if needed.